Block/Service/AbstractCategoriesBlockService.php

```diff
@@ -0,0 +1,188 @@
+    {
+        $adminField = $this->getFormAdminType($formMapper, $this->categoryAdmin, 'categoryId', 'category', array(
+            'label' => 'form.label_category',
+        ), array(
+            'translation_domain' => 'SonataClassificationBundle',
+            'link_parameters' => array(
+                array(
+                    array(
+                        'context' => $block->getSetting('context'),
+                    ),
+                ),
+            ),
+        ));
+
+        $formMapper->add('settings', 'sonata_type_immutable_array', array(
+            'keys' => array(
+                array('title', 'text', array(
+                    'label' => 'form.label_title',
+                    'required' => false,
+                )),
+                array('context', 'choice', array(
+                    'label' => 'form.label_context',
+                    'required' => false,
+                    'choices' => $this->getContextChoices(),
+                )),
+                array($adminField, null, array()),
+            ),
+            'translation_domain' => 'SonataClassificationBundle',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureSettings(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults(array(
+            'title' => 'Categories',
+            'category' => false,
+            'categoryId' => null,
+            'context' => 'default',
+            'template' => 'SonataClassificationBundle:Block:base_block_categories.html.twig',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function load(BlockInterface $block)
+    {
+        if (is_numeric($block->getSetting('categoryId'))) {
+            $block->setSetting('categoryId', $this->getCategory($block->getSetting('categoryId')));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function prePersist(BlockInterface $block)
+    {
+        $this->resolveIds($block);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function preUpdate(BlockInterface $block)
+    {
+        $this->resolveIds($block);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getBlockMetadata($code = null)
+    {
+        $description = (!is_null($code) ? $code : $this->getName());
+
+        return new Metadata($this->getName(), $description, false, 'SonataClassificationBundle', array(
+            'class' => 'fa fa-folder-open-o',
+        ));
+    }
+
+    /**
+     * @param mixed $id
+     * @param mixed $default
+     *
+     * @return CategoryInterface
+     */
+    final protected function getCategory($id, $default = false)
+    {
+        if (is_object($id)) {
+            return $id;
+        }
+
+        if (is_numeric($id)) {
+            return $this->categoryManager->find($id);
+        }
+
+        return $default;
+    }
+
+    /**
+     * @param BlockInterface $block
+     */
+    private function resolveIds(BlockInterface $block)
```

Comment:
> nobody could extend a private method

https://3v4l.org/oCG5c

> Why the hell should we make it final? Nobody knows the method (from the outside). It's no API method!

https://3v4l.org/oCG5c

Yes, it's horrifying. But this is technically possible.

In this case, make it final would avoid future issue like "You broke my ugly code making an ugly override on the last release!".